google/gax/api_callable.py

```diff
@@ -281,9 +288,68 @@ def construct_settings(
     # pylint: disable=too-many-locals
     defaults = {}
     bundle_descriptors = bundle_descriptors or {}
+    metrics_headers = metrics_headers or {}
     page_descriptors = page_descriptors or {}
     kwargs = kwargs or {}
 
+    # Initialize the headers, and add GAX and GRPC headers to the
+    # dictionary.
+    # pylint: disable=no-member
+    grpc_version = pkg_resources.get_distribution('grpcio').version
+    # pylint: enable=no-member
+    metrics_headers.update({
+        'gax': VERSION,
+        'gl-python': platform.python_version(),
+        'grpc': grpc_version,
+    })
+
+    # DEPRECATED: 2017-02-07
+    #       NOTE: Switch to loud warning when all new client libs are out.
+    #
+    # Sanity check: It is possible that we got this far but some headers
+    # were specified with an older library, which sends them as...
+    #   kwargs={'metadata': [('x-goog-api-client', 'foo/1.0 bar/3.0')]}
+    #
+    # Note: This is the final format we will send down to GRPC shortly.
+    #
+    # All new code sends this down as a dictionary at this point, but just
+    # in case we have intermingling things, try to read this and honor
+    # the stuff there if we get it.
+    #
+    # That said, if any values conflict, prefer the new-style ones.
+    if 'metadata' in kwargs:
+        for metadata in kwargs['metadata']:
+            # Sanity check: If this is a header other than x-goog-api-client,
+            # ignore it.
+            if metadata[0].lower() != 'x-goog-api-client':
+                continue
+
+            # Raise a warning, but be quiet about it (until we can update
+            # all the dependent libraries).
+            warnings.warn('Sending a x-goog-api-client header as metadata '
+                          'directly is deprecated; upgrade your dependent '
+                          'libraries.', PendingDeprecationWarning)
+
+            # Iterate over the space-separated thing/version identifiers
+            # and add any that we do not already have.
+            for ident in metadata[1].split(' '):
+                service, version = ident.split('/')
```

Comment:
Ditto. Only GAPIC has been using the `x-goog-api-client` header field. The hand-written layer has been using `user-agent`. No one outside Google should be using `x-goog-*`. Therefore, it's safer to just stomp on it, and emit a warning if anyone tries to pass in `x-goog-api-client` as part of `metadata`.